### ISSUE TITLE
Allow linking receipts to a transaction

### ISSUE THREAD

**Issue by AsBenDoge** [nlogozzo/NickvisionMoney#145]
Add receipt to a transaction

**Comment by nlogozzo** [comment]
Something like attaching an image/pdf? I like that idea.

**Comment by nlogozzo** [comment]
Would require a new column in the sql table of type BLOB. This will allow us to store images in the byte format directly in the database.

**Comment by nlogozzo** [comment]
![image](https://user-images.githubusercontent.com/17648453/210121385-3abfb5e6-392a-4b30-9fe8-f789ddb4f448.png)

This feature has been fully implemented in the backend and in the WinUI version. Going to work on the GNOME implementation now :) Just wanted to share progress!

**Comment by nlogozzo** [comment]
Just wanted to report back...we have the GNOME version almost working. There is just this weird bug where when you create a new transaction and upload a receipt to it, before you OK the dialog, you view the receipt and then OK the dialog...the application crashes. And we just get a bunch of gtk errors with no further explanation. Me and @fsobolev have been trying to figure it out for the past 3 hours but no luck. All the code has been committed to main @AsBenDoge if you'd like to try and investigate, but I believe this may be a gir.core issue in which we'd have to contain those developers, since this doesn't happen on Windows.

**Comment by AsBenDoge** [comment]
I'm sorry to hear that but I can't really help there. Fedora doesn't support .NET 7 yet so I can't build the application to test. Not to mention that I know nothing besides python... I could help with #3 pdf's template coordinates if you like my design, and that's probably the only code-related job I could do. Good luck to you guys :)

**Comment by badcel** [comment]
In case it helps: It is possible to install dotnet 7 for Linux manually in a user folder: https://dotnet.microsoft.com/en-us/download/dotnet/7.0

**Comment by nlogozzo** [comment]
> In case it helps: It is possible to install dotnet 7 for Linux manually in a user folder: https://dotnet.microsoft.com/en-us/download/dotnet/7.0

If you open the project in GNOME Builder, .NET 7 and other dependencies will be automatically downloaded and setup for you to run the app as well!

**Comment by fsobolev** [comment]
> If you open the project in GNOME Builder, .NET 7 and other dependencies will be automatically downloaded and setup for you to run the app as well!

Just make sure to configure project to use `*-gnomebuilder.json` manifest. The default one is for `flatpak-builder` and will fail in GNOME Builder.
The default manifest can be used like this:
```
flatpak run org.flatpak.Builder --user --install --force-clean _build org.nickvision.money.json
```

**Comment by nlogozzo** [comment]
> Just wanted to report back...we have the GNOME version almost working. There is just this weird bug where when you create a new transaction and upload a receipt to it, before you OK the dialog, you view the receipt and then OK the dialog...the application crashes. And we just get a bunch of gtk errors with no further explanation. Me and @fsobolev have been trying to figure it out for the past 3 hours but no luck. All the code has been committed to main @AsBenDoge if you'd like to try and investigate, but I believe this may be a gir.core issue in which we'd have to contain those developers, since this doesn't happen on Windows.

@badcel I need your help on this, I'm completely stuck. I'm trying to implement this feature in GNOME and it keeps crashing.

If you take a look at [TransactionDialog](https://github.com/nlogozzo/NickvisionMoney/blob/main/NickvisionMoney.GNOME/Views/TransactionDialog.cs) you'll see that I have an `Adw.ActionRow` with a `Gtk.MenuButton` and some `Gio.SimpleActions` for implementing the Receipt. These widgets are created as follows:
```csharp
//Group Receipt
_grpReceipt = Adw.PreferencesGroup.New();
_boxMain.Append(_grpReceipt);
//Receipt
var menuReceipt = Gio.Menu.New();
menuReceipt.Append(_controller.Localizer["View"], "transaction.receiptView");
menuReceipt.Append(_controller.Localizer["Delete"], "transaction.receiptDelete");
menuReceipt.Append(_controller.Localizer["Upload"], "transaction.receiptUpload");
_btnMenuReceipt = Gtk.MenuButton.New();
_btnMenuReceipt.SetValign(Gtk.Align.Center);
_btnMenuReceipt.AddCssClass("flat");
_btnMenuReceipt.SetLabel(_controller.Localizer["Manage"]);
_btnMenuReceipt.SetDirection(Gtk.ArrowType.Down);
_btnMenuReceipt.SetMenuModel(menuReceipt);
_rowReceipt = Adw.ActionRow.New();
_rowReceipt.SetTitle(_controller.Localizer["Receipt", "Field"]);
_rowReceipt.AddSuffix(_btnMenuReceipt);
_grpReceipt.Add(_rowReceipt);
//Receipt Action Map
_actMapReceipt = Gio.SimpleActionGroup.New();
_btnMenuReceipt.InsertActionGroup("transaction", _actMapReceipt);
//View Receipt Action
_actReceiptView = Gio.SimpleAction.New("receiptView", null);
_actReceiptView.OnActivate += (sender, e) => OnViewReceipt();
_actMapReceipt.AddAction(_actReceiptView);
//Delete Receipt Action
_actReceiptDelete = Gio.SimpleAction.New("receiptDelete", null);
_actReceiptDelete.OnActivate += (sender, e) => OnDeleteReceipt();
_actMapReceipt.AddAction(_actReceiptDelete);
//Upload Receipt Action
_actReceiptUpload = Gio.SimpleAction.New("receiptUpload", null);
_actReceiptUpload.OnActivate += (sender, e) => OnUploadReceipt();
_actMapReceipt.AddAction(_actReceiptUpload);
```

Each menu item is connected to the signals `OnViewReceipt()`, `OnDeleteReceipt()`, and `OnUploadReceipt()` respectively. 
Those functions are defined as follows:
```csharp
/// <summary>
/// Occurs when the view receipt button is clicked
/// </summary>
/// <param name="sender">Gtk.Button</param>
/// <param name="e">EventArgs</param>
private async void OnViewReceipt() => await _controller.OpenReceiptImageAsync(_receiptPath);

/// <summary>
/// Occurs when the delete receipt button is clicked
/// </summary>
/// <param name="sender">Gtk.Button</param>
/// <param name="e">EventArgs</param>
private void OnDeleteReceipt()
{
    _receiptPath = "";
    _actReceiptView.SetEnabled(false);
    _actReceiptDelete.SetEnabled(false);
}

/// <summary>
/// Occurs when the upload receipt button is clicked
/// </summary>
/// <param name="sender">Gtk.Button</param>
/// <param name="e">EventArgs</param>
private void OnUploadReceipt()
{
    var openFileDialog = Gtk.FileChooserNative.New(_controller.Localizer["Receipt", "Field"], _parentWindow, Gtk.FileChooserAction.Open, _controller.Localizer["Open"], _controller.Localizer["Cancel"]);
    openFileDialog.SetModal(true);
    var filterAll = Gtk.FileFilter.New();
    filterAll.SetName($"{_controller.Localizer["AllFiles"]} (*.jpg, *.jpeg, *.pdf)");
    filterAll.AddPattern("*.jpg");
    filterAll.AddPattern("*.jpeg");
    filterAll.AddPattern("*.pdf");
    openFileDialog.AddFilter(filterAll);
    var filterJpg = Gtk.FileFilter.New();
    filterJpg.SetName("JPG (*.jpg)");
    filterJpg.AddPattern("*.jpg");
    openFileDialog.AddFilter(filterJpg);
    var filterJpeg = Gtk.FileFilter.New();
    filterJpeg.SetName("JPEG (*.jpeg)");
    filterJpeg.AddPattern("*.jpeg");
    openFileDialog.AddFilter(filterJpeg);
    var filterPdf = Gtk.FileFilter.New();
    filterPdf.SetName("PDF (*.pdf)");
    filterPdf.AddPattern("*.pdf");
    openFileDialog.AddFilter(filterPdf);
    openFileDialog.OnResponse += (sender, e) =>
    {
        if (e.ResponseId == (int)Gtk.ResponseType.Accept)
        {
            var path = g_file_get_path(gtk_file_chooser_get_file(openFileDialog.Handle));
            _receiptPath = path;
            _actReceiptView.SetEnabled(true);
        }
    };
    openFileDialog.Show();
}
```
They are pretty basic methods.

Now, when I run the application and open the TransactionDialog, I click on the `Upload` menu item and select a file. 
Then, before I OK the dialog, I click on the `View` menu item (which opens the picture file with `xdg-open`). 
As soon as I press the `View` menu item, I get a bunch of GTK errors (but the application does not crash yet):
```bash
(NickvisionMoney.GNOME:661): GLib-GObject-CRITICAL **: 12:21:56.042: g_object_remove_toggle_ref: assertion 'G_IS_OBJECT (object)' failed
(NickvisionMoney.GNOME:661): GLib-GObject-CRITICAL **: 12:21:56.042: g_object_remove_toggle_ref: assertion 'G_IS_OBJECT (object)' failed
(NickvisionMoney.GNOME:661): GLib-GObject-CRITICAL **: 12:21:56.042: g_object_remove_toggle_ref: assertion 'G_IS_OBJECT (object)' failed
(NickvisionMoney.GNOME:661): GLib-GObject-CRITICAL **: 12:21:56.042: g_object_remove_toggle_ref: assertion 'G_IS_OBJECT (object)' failed
(NickvisionMoney.GNOME:661): GLib-GObject-CRITICAL **: 12:21:56.043: g_object_unref: assertion 'G_IS_OBJECT (object)' failed
(NickvisionMoney.GNOME:661): GLib-GObject-CRITICAL **: 12:21:56.043: g_object_unref: assertion 'G_IS_OBJECT (object)' failed
(NickvisionMoney.GNOME:661): GLib-GObject-CRITICAL **: 12:21:56.043: g_object_remove_toggle_ref: assertion 'G_IS_OBJECT (object)' failed
(NickvisionMoney.GNOME:661): GLib-GObject-CRITICAL **: 12:21:56.043: g_object_remove_toggle_ref: assertion 'G_IS_OBJECT (object)' failed
(NickvisionMoney.GNOME:661): GLib-GObject-CRITICAL **: 12:21:56.043: g_object_remove_toggle_ref: assertion 'G_IS_OBJECT (object)' failed
(NickvisionMoney.GNOME:661): GLib-GObject-CRITICAL **: 12:21:56.043: g_object_remove_toggle_ref: assertion 'G_IS_OBJECT (object)' failed
```
Then, when I go to close the dialog (either with Close or OK), the application crashes with the following error:
```bash
Process terminated. A callback was made on a garbage collected delegate of type 'NickvisionMoney.GNOME!NickvisionMoney.GNOME.Views.TransactionDialog+SignalCallback::Invoke'.
Repeat 2 times:
--------------------------------
   at Gio.Internal.Application.Run(IntPtr, Int32, System.String[])
--------------------------------
   at Gio.Application.Run()
   at NickvisionMoney.GNOME.Program.Run()
   at NickvisionMoney.GNOME.Program.Main(System.String[])
```
And @fsobolev and I can't figure out for the life of us why this is happening.

**Comment by nlogozzo** [comment]
> And @fsobolev and I can't figure out for the life of us why this is happening.

Fixed. Thanks @badcel <3
https://github.com/nlogozzo/NickvisionMoney/commit/e197ea1c4b96c7ac3012bb1e9d5c31098d5958c7

@fsobolev is going to double-test it on his end and send a screenshot :) 
I think I'm going to release beta2 once it's confirmed this works!

**Comment by nlogozzo** [comment]
![GNOME_Receipt](https://user-images.githubusercontent.com/17648453/210154627-c2d3681b-262f-47c2-831b-21caf6a3552a.png)

**Comment by nlogozzo** [comment]
Fixed in [V2023.1.0-beta2](https://github.com/nlogozzo/NickvisionMoney/releases/tag/2023.1.0-beta2)
The update is being published to flathub-beta now and should be available in 10-30 minutes :)
